ctmap: Introduce variable conntrack gc interval

tgraf opened the pull request:
Commit df8582d16e5 ("datapath: Optimize connection-tracking GC interval") has
introduced a regression by changing the default interval for LRU map based
conntrack to 12h on the basis that the tables will automatically expire the
oldest entries when filling up.

Unfortunately, walking the table to clean the conntrack stale podIPs on pod
removal taints all entries. This causes to random conntrack entries being
evicted while the table is full.

Resolve this by adjusting the conntrack gc interval automatically based on the
number of entries deleted in the table with the most deletions.

Fixes: #7997

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7999)
<!-- Reviewable:end -->

joestringer commented on pkg/endpointmanager/conntrack.go:
```diff
@@ -122,9 +112,10 @@ func EnableConntrackGC(ipv4, ipv6 bool, restoredEndpoints []*endpoint.Endpoint)
 
 	go func() {
 		for {
+			var maxDeleteRatio float64
```
Shouldn't this be outside the `for { ...`?

tgraf replied:
The delete ratio is calculated for each run. It's initialized there to make sure it is 0.0 for the per endpoint CT tables. That said, there was a check missing to skip the adjustment if the delete ratio is 0.0. I've added that.

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -550,3 +554,66 @@ func Exists(e CtEndpoint, ipv4, ipv6 bool) bool {
+		}
+		// 25%..90% => 1.3x..10x
+		interval = time.Duration(float64(interval) * (1.0 - maxDeleteRatio))
+		if interval > defaults.ConntrackGCMinInterval {
```
```suggestion
		if interval < defaults.ConntrackGCMinInterval {
```

tgraf replied:
Good catch, added test

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -550,3 +554,66 @@ func Exists(e CtEndpoint, ipv4, ipv6 bool) bool {
+		if maxDeleteRatio > 0.9 {
+			maxDeleteRatio = 0.9
+		}
+		// 25%..90% => 1.3x..10x
```
```suggestion
		// 25%..90% => 1.3x..10x shorter
```

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -550,3 +554,66 @@ func Exists(e CtEndpoint, ipv4, ipv6 bool) bool {
+		if maxDeleteRatio < 0.01 {
+			maxDeleteRatio = 0.01
+		}
+		// 5%..1% => 2x..10x
```
```suggestion
		// 5%..1% => 2x..10x longer
```

joestringer reviewed (changes requested):
I posted some minor feedback below. It's hard to tell how well these particular heuristic values will work in practice without some experience with them. I've opted not to think too deeply about them just yet.

A couple of questions for if this doesn't work as intended in some environments:

* Can we turn this off?
* Can we tweak the ratio thresholds/aggressiveness (even if such configuration is typically hidden from users)?

aanm commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -550,3 +554,72 @@ func Exists(e CtEndpoint, ipv4, ipv6 bool) bool {
+		}
+	}
+
+	if interval != prevInterval {
```
To avoid this condition from being true more times than necessary, round or truncate `interval` to the second, or even, to the minute. (Or switch the `Info` to `Debug`)

tgraf replied:
Very good point 👍

tgraf committed:
ctmap: Introduce variable conntrack gc interval

Commit df8582d16e5 ("datapath: Optimize connection-tracking GC interval") has
introduced a regression by changing the default interval for LRU map based
conntrack to 12h on the basis that the tables will automatically expire the
oldest entries when filling up.

Unfortunately, walking the table to clean the conntrack stale podIPs on pod
removal taints all entries. This causes to random conntrack entries being
evicted while the table is full.

Resolve this by adjusting the conntrack gc interval automatically based on the
number of entries deleted in the table with the most deletions.

Fixes: #7997

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Adjust documentation with new dynamic gc interval

Signed-off-by: Thomas Graf <thomas@cilium.io>
